calyx/src/passes/static_timing.rs

```diff
@@ -90,7 +91,8 @@ impl Visitor for StaticTiming {
                 let while_group_node =
                     st.insert_group(&while_group, HashMap::new())?;
 
-                let fsm_size = 32;
+                // `0` state + (2 + btime) states, where (2 + btime) == fsm_loop_exit_state
```

Comment:
Can you explain why you need the `2` here? What states does that correspond to?

Reply:
https://github.com/cucapra/futil/blob/e1eb074d0d2aa5f22b02ee5444323949e95dbda8/calyx/src/passes/static_timing.rs#L99

Reply:
Oh wow, I didn't even know there was this black magic. @sgpthomas needs to stare at this to make sure its correct.

Reply:
Yeah one thing I could do is declare these variables before `fsm_size` as well, so that we don't need to change code in two places. I'll wait for Sam to respond before I make any changes.

Reply:
For a body that takes n cycles, we use state 0 for the loop enter state, states 1 to n+1 for executing the body and n+2 for the loop exit state. It explains the state in the comment at the top of the while compilation function. Although I think there's an off by one typo in that. the correct comment should be:
```
                // FSM Encoding:
                //   0:     init state. we haven't started loop iterations
                //          and haven't checked the loop body
                //   1-n+1: body compute states. cond was true. compute the body
                //   n+2:   loop exit. we've finished running the body and the condition is false.
                // Transitions:
                //   0 -> 1:     when cond == true
                //   0 -> n+2:   when cond == false
                //   i -> i+1:   when i != 0 & i != n+1
                //   n+1-> 1:    when cond == true
                //   n+1 -> n+2: when cond == false
```

Reply:
jk, this is the correct one:
```
                // FSM Encoding:
                //   0:   init state. we haven't started loop iterations
                //        and haven't checked the loop body
                //   1-n: body compute states. cond was true. compute the body
                //   n+1: check condition again
                //   n+2: loop exit. we've finished running the body and the condition is false.
                // Transitions:
                //   0   -> 1:   when cond == true
                //   0   -> n+2: when cond == false
                //   i   -> i+1: when i != 0 & i != n
                //   n+1 -> 1:   when cond == true
                //   n+1 -> n+2: when cond == false
```

Reply:
Huh ok never saw that comment. I'll need to double check my numbers against this, and I'll update the comment as well. I'll resolve this when I'm done.